### PR TITLE
AOT compilation using ngtools/webpack

### PR DESCRIPTION
**Pull request by sendilkumarn**
- Please make sure the below checklist is followed for Pull Requests.

- [ ] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
- [x] Tests are added where necessary
- [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

File: generators/client/templates/angular/webpack/_webpack.prod.js [hunk 1]
```diff
@@ -36,6 +49,24 @@ module.exports = webpackMerge(commonConfig({ env: ENV }), {
         new Visualizer({
             // Webpack statistics in target folder
             filename: '../stats.html'
+        }),
+        // AOT Plugin 
+        new AotPlugin({
+            tsConfigPath: './tsconfig-aot.json',
+            entryModule: root('src/main/webapp/app/app.module#HipzeeAppModule')
```
Comment (deepu105) [inline generators/client/templates/angular/webpack/_webpack.prod.js]: hardcoded module name!

File: generators/client/templates/angular/webpack/_webpack.prod.js [hunk 2]
```diff
@@ -36,6 +49,24 @@ module.exports = webpackMerge(commonConfig({ env: ENV }), {
...
+        }),
+        new webpack.optimize.UglifyJsPlugin({
+            beautify: false,
+            comments: false,
+            sourcMap: true,
+            compress: {
+                screw_ie8: true,
```
Comment (deepu105) [inline generators/client/templates/angular/webpack/_webpack.prod.js]: :smiling_imp:

**Comment by sendilkumarn** [comment]
@deepu105 prod profile is working 
what is that minor mistake

**Comment by sendilkumarn** [comment]
except module name ?

**Comment by deepu105** [comment]
module name was the mistake i meant

File: generators/client/templates/angular/src/main/webapp/app/_app.main-aot.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+-%>
+import { platformBrowser } from '@angular/platform-browser';
+import { ProdConfig } from './blocks/config/prod.config';
+import { <%=angular2AppName%>AppModuleNgFactory } from '../../../../<%= BUILD_DIR %>aot/src/main/webapp/app/app.module.ngfactory';
```
Comment (deepu105) [inline generators/client/templates/angular/src/main/webapp/app/_app.main-aot.ts]: This might not be available all the time right?

**Comment by deepu105** [comment]
check failed build

File: generators/entity/templates/client/angular/src/main/webapp/app/entities/_entity.model.ts [hunk 4]
```diff
@@ -56,7 +56,7 @@ for (const idx in fields) {
         tsType = 'string';
     } else { //(fieldType === 'byte[]' || fieldType === 'ByteBuffer') && fieldTypeBlobContent == 'any' || (fieldType === 'byte[]' || fieldType === 'ByteBuffer') && fieldTypeBlobContent == 'image' || fieldType == 'LocalDate'
         tsType = 'any';
-        if ((fieldType === 'byte[]' || fieldType === 'ByteBuffer') && fields[idx].fieldTypeBlobContent == 'image') {
+        if ((fieldType === 'byte[]' || fieldType === 'ByteBuffer') && fields[idx].fieldTypeBlobContent !== 'text') {
```
Comment (sendilkumarn) [inline generators/entity/templates/client/angular/src/main/webapp/app/entities/_entity.model.ts]: Interesting, how this was not highlighted so far... 🤔

**Comment by deepu105** [comment]
There is also issue with the TS entity model class which didnt fail earlier
`Error at /home/deepu/workspace/temp/jh4test/target/aot/src/main/webapp/app/entities/product/product-dialog.component.ngfactory.ts:2434:45: Property 'subcategories' does not exist on type 'Product'.
`
When i changed `public subcategory?: SubCategory,` to `public subcategories?: SubCategory[],` it works

its coz of one to many relations declared in singular instead of plural as it appears in DTO or domain
so in `_entity.model.ts` line 65 to 76 needs to be altered. Now we need to check the relationship type and use proper plural form with type[] for DTO and no DTO cases. refer `_entity.java` and `entityDTO.java`

**Comment by deepu105** [comment]
try this fix 
```
relationships.forEach(relationship => {
    let fieldType;
    let fieldName;
    const relationshipType = relationship.relationshipType;
    if (relationshipType == 'one-to-many' || relationshipType == 'many-to-many') {
        fieldType = `${relationships.otherEntityAngularName}[]`;
        fieldName = relationships.relationshipFieldNamePlural;
    } else {
        if (dto == "no") {
            fieldType = relationships.otherEntityAngularName;
            fieldName = relationships.relationshipFieldName;
        } else {
            fieldType = tsKeyType;
            fieldName = relationships.relationshipFieldName + "Id";
        }
    }
    variables[fieldName] = fieldName + '?: ' + fieldType;
});
```

could also convert the for loop on fields on same file to a forEach?

**Comment by deepu105** [comment]
ok this fix lets it compile but still I get no screens. I see below error @sendilkumarn can you test it on a clean fresh project?
```

d66386d….main.bundle.js:1 ERROR Error: Uncaught (in promise): Error: No provider for CookieOptions!
Error: No provider for CookieOptions!
    at injectionError (d66386d….main.bundle.js:1)
    at noProviderError (d66386d….main.bundle.js:1)
    at ReflectiveInjector_._throwOrNull (d66386d….main.bundle.js:1)
    at ReflectiveInjector_._getByKeyDefault (d66386d….main.bundle.js:1)
    at ReflectiveInjector_._getByKey (d66386d….main.bundle.js:1)
    at ReflectiveInjector_.get (d66386d….main.bundle.js:1)
    at Jh4TestAppModuleInjector.get [as _CookieService_135] (d66386d….main.bundle.js:1)
    at Jh4TestAppModuleInjector.get [as _CSRFService_138] (d66386d….main.bundle.js:1)
    at Jh4TestAppModuleInjector.get [as _JhiTrackerService_139] (d66386d….main.bundle.js:1)
    at Jh4TestAppModuleInjector.get [as _Principal_140] (d66386d….main.bundle.js:1)
    at injectionError (d66386d….main.bundle.js:1)
    at noProviderError (d66386d….main.bundle.js:1)
    at ReflectiveInjector_._throwOrNull (d66386d….main.bundle.js:1)
    at ReflectiveInjector_._getByKeyDefault (d66386d….main.bundle.js:1)
    at ReflectiveInjector_._getByKey (d66386d….main.bundle.js:1)
    at ReflectiveInjector_.get (d66386d….main.bundle.js:1)
    at Jh4TestAppModuleInjector.get [as _CookieService_135] (d66386d….main.bundle.js:1)
    at Jh4TestAppModuleInjector.get [as _CSRFService_138] (d66386d….main.bundle.js:1)
    at Jh4TestAppModuleInjector.get [as _JhiTrackerService_139] (d66386d….main.bundle.js:1)
    at Jh4TestAppModuleInjector.get [as _Principal_140] (d66386d….main.bundle.js:1)
    at resolvePromise (d66386d….polyfills.bundle.js:1)
    at resolvePromise (d66386d….polyfills.bundle.js:1)
    at d66386d….polyfills.bundle.js:1
    at ZoneDelegate.invokeTask (d66386d….polyfills.bundle.js:1)
    at Object.onInvokeTask (d66386d….main.bundle.js:1)
    at ZoneDelegate.invokeTask (d66386d….polyfills.bundle.js:1)
    at Zone.runTask (d66386d….polyfills.bundle.js:1)
    at drainMicroTaskQueue (d66386d….polyfills.bundle.js:1)
    at <anonymous>
```

File: generators/entity/templates/client/angular/src/main/webapp/app/entities/_entity.model.ts [hunk 5]
```diff
@@ -56,24 +56,30 @@ for (const idx in fields) {
...
-        fieldType = relationships[idx].otherEntityAngularName;
-        fieldName = relationships[idx].relationshipFieldName;
+    const relationshipType = relationship.relationshipType;
+    if (relationshipType == 'one-to-many' || relationshipType == 'many-to-many') {
+        fieldType = `${relationships.otherEntityAngularName}[]`;
+        fieldName = relationships.relationshipFieldNamePlural;
     } else {
-        fieldType = tsKeyType;
-        fieldName = relationships[idx].relationshipFieldName + "Id";
+        if (dto == "no") {
```
Comment (deepu105) [inline generators/entity/templates/client/angular/src/main/webapp/app/entities/_entity.model.ts]: ahh `dto === 'no'` plz i missed it

Comment (deepu105) [inline generators/entity/templates/client/angular/src/main/webapp/app/entities/_entity.model.ts]: @sendilkumarn

**Comment by deepu105** [comment]
protractor tests still failing probably same error as above when running in prod profile

**Comment by deepu105** [comment]
btw also rebase with current master

**Comment by deepu105** [comment]
@sendilkumarn are you stuck? do you need help? we really need this to proceed with #4794

File: generators/client/templates/angular/webpack/_webpack.prod.js [hunk 6]
```diff
@@ -38,6 +51,24 @@ module.exports = webpackMerge(commonConfig({ env: ENV }), {
...
+        }),
+        // AOT Plugin 
+        new AotPlugin({
+            tsConfigPath: './tsconfig-aot.json',
+            entryModule: root('src/main/webapp/app/app.module#<%=angular2AppName%>AppModule')
+        }),
+        new webpack.optimize.UglifyJsPlugin({
+            beautify: false,
+            comments: false,
+            sourcMap: true,
```
Comment (thomas-burko) [inline generators/client/templates/angular/webpack/_webpack.prod.js]: typo?

**Comment by deepu105** [comment]
@sendilkumarn @jdubois I had to revert this PR as its causing some sort of infinite loop in ngc phase

**Comment by deepu105** [comment]
I'll create a new branch and PR to work on this

**Pull request merged**